crates/sdk/src/client_cache.rs

```diff
@@ -112,41 +116,186 @@ impl<M: SpacetimeModule> ClientCache<M> {
             return;
         }
 
-        // Clippy is incorrect here: `.cloned` will do `Arc::clone`, not `TableCache::clone`.
-        // TODO: Do we need to be `Arc`ing these? `im::HashMap` should be doing sharing internally anyway.
-        #[allow(clippy::map_clone)]
-        let mut table = self
-            .get_table::<Row>(table_name)
-            .map(|tbl| TableCache::clone(tbl))
-            .unwrap_or_default();
+        let table = self.get_or_make_table::<Row>(table_name);
 
         table.apply_diff(diff);
-        self.tables
-            .entry::<HashMap<&'static str, Arc<TableCache<Row>>>>()
-            .or_insert(HashMap::default())
-            .insert(table_name, Arc::new(table));
     }
 }
 
-/// A shared view into a particular state of the `ClientCache`.
-pub(crate) type ClientCacheView<M> = Arc<ClientCache<M>>;
+/// Internal implementation of a generated `TableHandle` struct,
+/// which mediates access to a table in the client cache.
+///
+/// `TableHandle`s don't actually hold a direct reference to the table they access,
+/// as that would require both gnarly lifetimes and also a `MutexGuard` on the client cache.
+/// Instead, they hold an `Arc<Mutex>` on the whole [`ClientCache`],
```

Comment:
obligatory:
- what if we used `RwLock`
- nah we can do that later.

Reply:
That doesn't fix the "gnarly lifetimes" issue. And it actually seems likely to introduce *more* deadlocks, not less, as in that version we would have long-standing read guards on the client cache, whereas in this version we have only short-lived exclusive guards.